### PR TITLE
Adding compiler warning

### PR DESCRIPTION
**Pull request by SebaLukas**
## Describe your changes
Adding gcc compiler warnings `Wall, Wextra & Wno-unused-function` to libiso15118. At the same time, the warnings that occur are fixed here.

## Issue ticket number and link
Right now there was no compiler warnings enabled. This PR should fix this.

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

File: src/iso15118/session/iso.cpp [hunk 1]
```diff
@@ -157,14 +157,18 @@ TimePoint const& Session::poll() {
     }
 
     // send all of our queued control events
-    while (active_control_event = control_event_queue.pop()) {
+    active_control_event =
+        control_event_queue.pop(); // TODO(sl): Not sure, -Wparentheses does not like the previous solution
```
Comment (barsnick) [inline src/iso15118/session/iso.cpp:161]: To assign with a conditional check, the assignment can be put into additional brackets:
```
while ((active_control_event = control_event_queue.pop())) {
```

While it looks like a hack to shut up the warnings, it's apparently the way to tell the compiler about your intent.

Comment (a-w50) [reply src/iso15118/session/iso.cpp:161]: https://stackoverflow.com/questions/24116817/when-do-extra-parentheses-have-an-effect-other-than-on-operator-precedence/24116818#24116818

additional parentheses can have a real subtle effect

File: src/iso15118/io/sdp_server.cpp [hunk 2]
```diff
@@ -138,7 +138,7 @@ TlsKeyLoggingServer::TlsKeyLoggingServer(const std::string& interface_name, uint
     auto could_bind = false;
     auto source_port = 49152;
     for (; source_port < 65535; source_port++) {
-        sockaddr_in6 source_address = {AF_INET6, htons(source_port)};
```
Comment (a-w50) [inline src/iso15118/io/sdp_server.cpp:141]: was this by accident, that `source_port` got replaced by `port`?

File: src/iso15118/io/connection_openssl.cpp [hunk 3]
```diff
@@ -107,7 +107,7 @@ static SSL_CTX* init_ssl(const config::SSLConfig& ssl_config) {
                 return;
             }
             const auto result = key_logging_server.send(line);
-            if (result != strlen(line)) {
+            if (result < 0 or (result >= 0 and static_cast<size_t>(result) != strlen(line))) {
```
Comment (a-w50) [inline src/iso15118/io/connection_openssl.cpp:110]: `result < 0 or (static_cast<size_t>(result) != strlen(line))` should be enough I guess.  If this pattern is used far more often, we could also define a helper template like
```
template <typename Signed, typename Unsigned>
constexpr bool safe_compare(Signed signed_value, Unsigned unsigned_value) noexcept {
    static_assert(std::is_signed_v<Signed>, "signed parameter needs to be signed type");
    static_assert(std::is_unsigned_v<Unsigned>, "unsigned parameter needs to be unsigned type");

    if (signed_value < 0) {
        return false;
    }

    if (signed_value > std::numeric_limits<Unsigned>::max()) {
        return false;
    }

    return static_cast<Unsigned>(signed_value) == unsigned_value;
}
```
Probably it is not generic enough and has some bugs, but you get the idea

File: src/iso15118/io/sdp_server.cpp (hunk 2 above)
Comment (SebaLukas) [reply src/iso15118/io/sdp_server.cpp:141]: Yes, thx for pointing that out :+1:

File: src/iso15118/io/connection_openssl.cpp (hunk 3 above)
Comment (a-w50) [reply src/iso15118/io/connection_openssl.cpp:110]: who would have known:
https://en.cppreference.com/w/cpp/utility/intcmp
so this should be the way to go:
```
template<class T, class U>
constexpr bool cmp_equal(T t, U u) noexcept
{
    if constexpr (std::is_signed_v<T> == std::is_signed_v<U>)
        return t == u;
    else if constexpr (std::is_signed_v<T>)
        return t >= 0 && std::make_unsigned_t<T>(t) == u;
    else
        return u >= 0 && std::make_unsigned_t<U>(u) == t;
}
```

File: src/iso15118/session/iso.cpp (hunk 1 above)
Comment (a-w50) [reply src/iso15118/session/iso.cpp:161]: but here, they are used to express intend

Comment (SebaLukas) [reply src/iso15118/session/iso.cpp]: Done

File: src/iso15118/io/sdp_server.cpp (hunk 2 above)
Comment (SebaLukas) [reply src/iso15118/io/sdp_server.cpp:141]: Done

File: src/iso15118/io/connection_openssl.cpp (hunk 3 above)
Comment (SebaLukas) [reply src/iso15118/io/connection_openssl.cpp]: Done, thx for finding `intcmp`!

File: src/iso15118/io/connection_openssl.cpp [hunk 4]
```diff
@@ -107,7 +107,7 @@ static SSL_CTX* init_ssl(const config::SSLConfig& ssl_config) {
                 return;
             }
             const auto result = key_logging_server.send(line);
-            if (result < 0 or (result >= 0 and static_cast<size_t>(result) != strlen(line))) {
+            if (result < 0 or cmp_equal(result, strlen(line))) {
```
Comment (a-w50) [inline src/iso15118/io/connection_openssl.cpp:110]: shouldn't this be just `if (not cmp_equal(result, strlen(line)))`?

Comment (SebaLukas) [reply src/iso15118/io/connection_openssl.cpp:110]: I think `result` will be `-1` if an error has occurred during `send()`. And the cmp_equal does not check that

Comment (a-w50) [reply src/iso15118/io/connection_openssl.cpp:110]: yes, but still `-1` is not equal to `strlen(line)`

Comment (SebaLukas) [reply src/iso15118/io/connection_openssl.cpp:110]: Yeaah you are correct. I'll change that

**Comment by SebaLukas** [comment]
Setting the compiler warnings as default and as errors is actually not a bad idea :+1:

**Comment by SebaLukas** [comment]
Using `target_compile_options` instead of `target_compile_definitions`

**Review by a-w50 — APPROVED** [review]
all fine!

**Pull request merged**